returnn/hdf.py

```diff
@@ -346,6 +364,296 @@ def run(self):
+            # stalling the worker processes (if we don't).
+            (rec.audio, segments)
+            for rec in c.all_recordings()
+            if len(
+                segments := [
+                    (full_name, segment.start, segment.end)
+                    for segment in rec.segments
+                    if (full_name := segment.fullname()) in all_whitelists
+                ]
+            )
+            > 0
+        )
+        # use imap instead of imap_unordered to have reproducible results, even though this
+        # causes a slight speed penalty
+        for results in pool.imap(self._process_seq, recs, chunksize=8):
+            for (segment_name, data), (hdf_writer, segments_whitelist) in itertools.product(
+                results, zip(hdf_writers, segment_whitelists)
+            ):
+                if segment_name not in segments_whitelist:
+                    continue
+                _logging.info(f"Writing {segment_name} to {hdf_writer.filename}.")
+                hdf_writer.insert_batch(
+                    inputs=data.reshape(1, -1, 1),
+                    seq_len=[data.shape[0]],
+                    seq_tag=[segment_name],
+                )
+        for hdf_writer in hdf_writers:
+            hdf_writer.close()
+
+    def _process_seq(self, recording: Tuple[str, Sequence[Tuple[str, float, float]]]) -> List[Tuple[str, np.ndarray]]:
+        audio_file, segments = recording
+
+        assert isinstance(self.multi_channel_strategy, (BlissToPcmHDFJob.Mixdown, BlissToPcmHDFJob.PickNth)), (
+            "unknown multi_channel_strategy"
+        )
+        channel_mix_args = (
+            ["-ac", "1"]
+            if isinstance(self.multi_channel_strategy, BlissToPcmHDFJob.Mixdown)
+            else ["-map_channel", f"0.{self.multi_channel_strategy.channel}"]
+        )
+
+        # We preprocess all data with ffmpeg because it is more robust to different
+        # audio formats like gsm-ms (which soundfile cannot read) or mp3 (standard format,
+        # but soundfile cannot seek). We also resample it at the same time and merge channels.
+        #
+        # We then use the wave library to extract the segments because we want to slice the audio
+        # data on a frame-by-frame basis (to match with RASR), while ffmpeg supports only
+        # temporal slices.
+        ffmpeg_proc = sp.run(
+            [
+                "ffmpeg",
+                "-hide_banner",
+                "-loglevel",
+                "error",
+                "-y",
+                "-i",
+                audio_file,
+                "-c:a",
+                "pcm_s16le",
+            ]
+            + channel_mix_args
+            + [
+                "-af",
+                "aresample=resampler=soxr",
+                "-ar",
+                str(self.target_sampling_rate),
+                "-f",
+                "wav",
+                "-",
+            ],
+            check=True,
+            stdout=sp.PIPE,
+        )
```

Comment:
remove training comma to let black collapse the lists:
```suggestion
        ffmpeg_proc = sp.run(
            ["ffmpeg", "-hide_banner", "-loglevel", "error", "-y", "-i", audio_file, "-c:a", "pcm_s16le"]
            + channel_mix_args
            + ["-af", "aresample=resampler=soxr", "-ar", str(self.target_sampling_rate), "-f", "wav", "-"],
            check=True,
            stdout=sp.PIPE,
        )
```